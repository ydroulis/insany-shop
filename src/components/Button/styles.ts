'use client'
import styled from 'styled-components'

export const Container = styled.button<{ $buttonType: string }>`
    display: flex;
    gap: 16px;
    justify-content: center;
    align-items: center;

    border: none;
    border-radius: 4px;

    background-color: ${({ $buttonType }) => ($buttonType === 'default' ? '#000000' : '#1B9847')};
    cursor: pointer;
    width: 100%;
    height: ${({ $buttonType }) => ($buttonType === 'default' ? '40px' : '44px')};

    color: ${({ $buttonType }) => ($buttonType === 'default' ? '#fff' : '#f5f5fa')};
    font-size: 1rem;
    font-weight: 500;
    text-transform: ${({ $buttonType }) => ($buttonType === 'default' ? 'none' : 'uppercase')};

    p{
        margin: 0;
    }

    &:hover {
        opacity: 0.8;
    }

    &:focus-visible {
        outline: 3px solid #2563eb;
        outline-offset: 4px;
        border-radius: 4px;
        opacity: 0.8;
    }
`