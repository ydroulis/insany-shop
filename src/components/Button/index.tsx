import React from 'react';
import { FiShoppingCart } from "react-icons/fi";

import * as S from './styles';

interface ButtonProps {
    children: React.ReactNode;
    type?: 'default' | 'checkout';
    action: () => void;
    ariaLabel?: string
}

const Button: React.FC<ButtonProps> = ({ children, type = "default", action, ariaLabel }) => {
    const accessibleLabel = ariaLabel ?? String(children);

    return (
        <S.Container $buttonType={type} onClick={() => action()} type='button' aria-label={accessibleLabel}>
            {type === "default" && <FiShoppingCart data-testid='icon-svg' size={24} color='#fff' aria-hidden="true" focusable="false" />}
            <p>{children}</p>
        </S.Container>
    );
}

export default Button;