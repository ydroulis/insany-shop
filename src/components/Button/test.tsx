import { render, screen, fireEvent } from "@testing-library/react";
import Button from "./";

describe("<Button />", () => {
    it("should render with children text", () => {
        render(<Button action={jest.fn()}>Comprar</Button>);

        const button = screen.getByRole("button", { name: /comprar/i });
        expect(button).toBeInTheDocument();
        expect(button).toHaveTextContent("Comprar");
    });

    it("should call action on click", () => {
        const action = jest.fn();
        render(<Button action={action}>Adicionar</Button>);

        const button = screen.getByRole("button", { name: /adicionar/i });
        fireEvent.click(button);

        expect(action).toHaveBeenCalledTimes(1);
    });

    it("should use aria-label when provided", () => {
        render(<Button action={jest.fn()} ariaLabel="Adicionar ao carrinho">+</Button>);

        const button = screen.getByRole("button", { name: /adicionar ao carrinho/i });
        expect(button).toBeInTheDocument();
        expect(button).toHaveAttribute("aria-label", "Adicionar ao carrinho");
    });

    it("should render shopping cart icon when type is default", () => {
        render(<Button action={jest.fn()}>Cart</Button>);

        const icon = screen.getByTestId("icon-svg"); // vamos configurar isso abaixo
        expect(icon).toBeInTheDocument();
        expect(icon).toHaveAttribute("aria-hidden", "true");
    });

    it("should not render shopping cart icon when type is checkout", () => {
        render(<Button type="checkout" action={jest.fn()}>Finalizar</Button>);

        const button = screen.getByRole("button", { name: /finalizar/i });
        expect(button).toBeInTheDocument();

        expect(button.querySelector("svg")).toBeNull();
    });
    it("should match snapshot", () => {
        const { container } = render(<Button action={jest.fn()}>Comprar</Button>);
        expect(container).toMatchSnapshot();
    });
});
