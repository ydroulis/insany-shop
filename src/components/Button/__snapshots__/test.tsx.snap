// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Button /> should match snapshot 1`] = `
<div>
  <button
    aria-label="Comprar"
    class="sc-bRKDuR iGlgTQ"
    type="button"
  >
    <svg
      aria-hidden="true"
      color="#fff"
      data-testid="icon-svg"
      fill="none"
      focusable="false"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="color: rgb(255, 255, 255);"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="9"
        cy="21"
        r="1"
      />
      <circle
        cx="20"
        cy="21"
        r="1"
      />
      <path
        d="M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6"
      />
    </svg>
    <p>
      Comprar
    </p>
  </button>
</div>
`;
